File: ios/Classes/api/OneToOneCallApi.swift
import JCSDKOC

import Foundation

class OneToOneCallApi : JyphoonCallApi {
    
    func call(destination: String, password: String, video: Bool, did: String, type: CallType, ts: Int32) -> Bool {
        let callParam = JCCallParam(
            extraParam: video ? "video" : "audio",
            ticket: "\(did)_\(ts)"
        )
        JCRoom.shared.call.mediaConfig = JCCallMediaConfig.generate(by: JCCallMediaConfigMode.modeIOT)
        let res = JCRoom.shared.call.call(
            destination,
            video: video,
            callParam: callParam
        )
        if (!res) {
            return false;
        }
        setAudio(audio: true)
        if (video) {
            setVideo(video: true)
            setSpeaker(speaker: true)
        }
        
        return res
    }
    
    func speaker() -> Bool {
        JCRoom.shared.mediaDevice.isSpeakerOn()
    }
    
    func callStatus() -> String {
        guard let activeCallItem = JCRoom.shared.call.getActiveCallItem() else {
            return "off"
        }
        
        switch (activeCallItem.state) {
        case .`init`, .pending, .connecting:
            return "waiting"
        case .talking, .ok:
            return "on"
        case .cancel, .canceled, .missed, .error:
            return "off"
        @unknown default:
            return "off"
        }
    }
    
    func audio() -> Bool {
        let audioStart = JCRoom.shared.mediaDevice.audioStart
        let uploadLocalAudio = JCRoom.shared.call.getActiveCallItem()?.microphoneMute ?? true
        return audioStart && !uploadLocalAudio
    }
    
    func video() -> Bool {
        let videoStart = JCRoom.shared.mediaDevice.cameraOpen
        let uploadVideoStreamSelf = JCRoom.shared.call.getActiveCallItem()?.uploadVideoStreamSelf ?? false
        let video = JCRoom.shared.call.getActiveCallItem()?.video ?? false
        return videoStart && uploadVideoStreamSelf && video
    }
    
    func otherAudio() -> Bool {
        return !(JCRoom.shared.call.getActiveCallItem()?.otherAudioInterrupt ?? false)
    }
    
    func otherVideo() -> Bool {
        return !(JCRoom.shared.call.getActiveCallItem()?.uploadVideoStreamOther ?? false)
    }
    
    func leave() -> Bool {
        let call = JCRoom.shared.call.getActiveCallItem()
        if (call == nil) {
            return false
        }
        return JCRoom.shared.call.term(call!, reason: JCCallReason.none, description: "")
    }
    
    func switchCamera() {
        JCRoom.shared.mediaDevice.switchCamera()
    }
    
    func setVideo(video: Bool) {
        if (video) {
            JCRoom.shared.mediaDevice.startCamera()
        } else {
            JCRoom.shared.mediaDevice.stopCamera()
        }
    }
    
    func setAudio(audio: Bool) {
        if (audio) {
            JCRoom.shared.mediaDevice.startAudio()
        } else {
            JCRoom.shared.mediaDevice.stopAudio()
        }
        JCHandler.instance.onEvent(event: JCHandler.CALL_UPDATE)
    }
    
    func setSpeaker(speaker: Bool) {
        JCRoom.shared.mediaDevice.enableSpeaker(speaker)
        JCHandler.instance.onEvent(event: JCHandler.CALL_UPDATE)
    }
}

File: android/src/main/kotlin/com/michaellazebny/jyphoon/jc/api/OneToOneCallApi.kt
package com.michaellazebny.jyphoon.jc.api

import android.util.Log
import com.juphoon.cloud.JCCall
import com.michaellazebny.jyphoon.jc.CallType
import com.michaellazebny.jyphoon.jc.JyphoonCallApi
import com.michaellazebny.jyphoon.jc.jcWrapper.JCManager

class OneToOneCallApi : JyphoonCallApi {
    /**
     * Starts the "call".
     * Returns true if the call was started successfully.
     *
     * [destination] - the identifier of the companion or the conference.
     *
     * [password] - the password for the conference.
     *
     * [video] - initiate call as videocall or audiocall.
     *
     * [type] - oneToOne or group.
     */
    override fun call(
        destination: String,
        password: String,
        video: Boolean,
        did: String,
        type: CallType,
        ts: Long
    ): Boolean {
        val param = JCCall.CallParam(
            if (video) "video" else "audio",
            "${did}_$ts",
        )
        val res = JCManager.getInstance().call.call(destination, video, param)
        if (res && video) {
            setVideo(true)
        }
        return res
    }

    /**
     * Returns CallStatus.
     * Returns the current call status.
     * It can be one of the following:
     * - CallStatus.on
     * - CallStatus.off
     * - CallStatus.waiting
     */
    override fun callStatus(): String {
        val call = JCManager.getInstance().call
        val activeCall = call.activeCallItem ?: return "off"
        JCCall.REASON_NONE
        Log.i("JC_CALL_STATUS_REASON", activeCall.reason.toString())

        return when (activeCall.state) {
            JCCall.STATE_OK, JCCall.STATE_TALKING -> "on"
            JCCall.STATE_PENDING, JCCall.STATE_CONNECTING, JCCall.STATE_INIT -> "waiting"
            else -> "off"
        }
    }

    /**
     * True if the user is in the call and is not muted.
     * Otherwise, returns false.
     */
    override fun audio(): Boolean {
        val audioStart = JCManager.getInstance().mediaDevice.isAudioStart
        val uploadLocalAudio = JCManager.getInstance().call.activeCallItem?.microphoneMute ?: false
        return audioStart && uploadLocalAudio
    }

    /**
     * True if the user is in the call and shares video.
     * Otherwise, returns false.
     */
    override fun video(): Boolean {
        val videoStart = JCManager.getInstance().mediaDevice.isCameraOpen
        val uploadVideoStreamSelf =
            JCManager.getInstance().call.activeCallItem.uploadVideoStreamSelf
        val video = JCManager.getInstance().call.activeCallItem.video
        return videoStart && uploadVideoStreamSelf && video
    }

    /**
     * True if the companion is in the call and is not muted.
     * Otherwise, returns false.
     */
    override fun otherAudio(): Boolean {
        return !JCManager.getInstance().call.activeCallItem.otherAudioInterrupt
    }

    /**
     * True if the companion is in the call and shares video.
     * Otherwise, returns false.
     */
    override fun otherVideo(): Boolean {
        return JCManager.getInstance().call.activeCallItem.uploadVideoStreamOther
    }

    /**
     * Hangs up the call or leaves the conference
     * depending on the call type.
     */
    override fun leave(): Boolean {
        val call = JCManager.getInstance().call
        call.term(call.activeCallItem, JCCall.REASON_NONE, "term")
        return true
    }

    /** Switches the camera (front/back) */
    override fun switchCamera() {
        JCManager.getInstance().mediaDevice.switchCamera()
    }

    /** Enables \ disables video. */
    override fun setVideo(video: Boolean) {
        if (video) {
            JCManager.getInstance().mediaDevice.startCamera()
        } else {
            JCManager.getInstance().mediaDevice.stopCamera()
        }
    }

    /** Enables \ disables audio. */
    override fun setAudio(audio: Boolean) {
        if (audio) {
            JCManager.getInstance().mediaDevice.startAudio()
        } else {
            JCManager.getInstance().mediaDevice.stopAudio()
        }
    }

    /** Enables \ disables speaker. */
    override fun setSpeaker(speaker: Boolean) {
        JCManager.getInstance().mediaDevice.enableSpeaker(speaker)
    }

}